### DISCUSSION TITLE
File Logging

### DISCUSSION THREAD

**Discussion by DGuhr** [keycloak/keycloak#10497] [Archive]
### Current situation:
We're relying on unsupported raw quarkus properties if you want to use Keycloak 17 and file logging. That should not stay this way, bc. file logging is still somewhat essential for a subset of environments, so we should support it. First usecases were pointed out in #8870 by @trixpan and @cgeorgilakis 

### Proposal:

Add the following config properties to keycloak directly, matching ones can be found in https://quarkus.io/guides/logging#quarkus-log-logging-log-config_quarkus.log.file-file-logging

### Notation 
Config options are shown as cli options for better readability, but would also be available in KC_ notation as env variables or without the prefix `--` in `keycloak.conf`
 
### Issue / Milestone 1

`--log-file-enabled=<bool>`

`--log-file-level=<log-level>`

`--log-file=<path>/<filename>:data/log/keycloak.log` 

`--log-file-format=<format>` 

### Issue / Milestone 2

`--log-file-rotate-max-file-size=<MemorySize>`-  byte to yottabytes is generally possible by quarkus, imo from kb to ...terra? should suffice?> format: `<number>`B,K,M,G,T, so e.g. `size=100k` would create a new file when 100kb are exceeded.

`--log-file-rotate-max-backup-index=<number of files to keep before deleting oldest>` (not 100% sure if we can't come up with better naming, e.g. `files-to-keep` or sth) 

`--log-file-rotate-on-boot-enabled=<true/false>` (only available when at least max-file-size or file-suffix is set)

### Milestone 2 or 3 (unclear atm, afaik would allow the usecase to auto-delete files after xx days with an additional script )
`--log-file-rotation-file-suffix=<suffix>` - looking at the quarkus code it could be zip, gz or a date pattern, but have to check further while implementing. Seems its using jboss logmanager capabilitiesdirectly, especially `PeriodicSizeRotatingFileHandler` and `PeriodicRotatingFileHandler` - file in quarkus is LoggingSetupRecorder`

Considerations:
- see comments above
- default path and filename: data/server.log? => data/log/keycloak.log it is for now,
- does the filename need to end with .log, or is everything allowed? should we not allow everything? => everything allowed.
- see file-suffix.
- logging alone will add a huge amount of config properties, the full help page at least in the cli will get a bit messy. We had several thoughts of something like an "expert mode", or to be able to filter configuration options by category (or even by key). Will open another discussion for this. -> #10499 
- allow async logging? (see quarkus config reference) what would the use cases be? => no support.
- directory as value for `--log-file=<path>/<filename>` results in error (not meaning kc does not start).

**Comment by DGuhr** [comment]
Initialfile logging support is now in latest main, see #10576 - expect some more options when we get to #10618 - we need quarkus 2.8.0 to be released though.

**Comment by EricGoj** [comment]
Hello @DGuhr , can you help with this issue?

I'm trying to save the Keycloak file loggin in a directory within my cloud. I'm running Keycloak using docker-compose, and my docker-compose.yml looks like this:
` keycloak-service:
    image: quay.io/keycloak/keycloak:20.0.0
    ports:
      - '8989:8080'
    env_file:
      - ./keycloak/variables.txt
    volumes:
      - type: bind
        source: ./keycloak/themes/custom
        target: /opt/keycloak/themes/custom
      - type: bind
        source: ./keycloak/email-2fa-app/keycloak-2fa-email-15.jar
        target: /opt/keycloak/providers/keycloak-2fa-email-15.jar
      - type: bind
        source: ./keycloak/logs
        target: /opt/keycloak/data/log
    environment:
      - TZ=America/Argentina/Buenos_Aires
    command: start --http-relative-path=/auth --log=console,file --log-file=/keycloak/logs/kc.log`

But when the container starts, it shows the following error:

`keycloak-service-1  | LogManager error of type OPEN_FAILURE: Failed to set log file
keycloak-service-1  | java.io.FileNotFoundException: ./keycloak/logs/kc.log (No such file or directory)
keycloak-service-1  |   at java.base/java.io.FileOutputStream.open0(Native Method)
keycloak-service-1  |   at java.base/java.io.FileOutputStream.open(FileOutputStream.java:298)
keycloak-service-1  |   at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:237)
keycloak-service-1  |   at org.jboss.logmanager.handlers.FileHandler.setFile(FileHandler.java:148)
keycloak-service-1  |   at org.jboss.logmanager.handlers.SizeRotatingFileHandler.setFileInternal(SizeRotatingFileHandler.java:252)`

I'm using Keycloak version 20.0.0 default, and I haven't made any changes. I would greatly appreciate a response. Thank you!
